test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,289 @@
+// small size often uses single-work-group specialization for device policy
+// large size often uses multi-work-group specialization
+constexpr std::size_t large_size = 81207;
+constexpr std::size_t small_size = 4134;
+
+template<typename KeyIt, typename ValIt, typename Size>
+Size
+remove_duplicates_by_key(const KeyIt& keys_begin, const ValIt& vals_begin, Size n)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    std::unordered_map<KeyT, Size> histogram;
+    std::for_each(keys_begin, keys_begin + n, [&histogram](const auto& key) { ++histogram[key]; });
+    auto first = oneapi::dpl::make_zip_iterator(keys_begin, vals_begin);
+    auto has_duplicates = [&histogram](const auto& pair) {return histogram[std::get<0>(pair)] > 1; };
+    auto new_last = std::remove_if(first, first + n, has_duplicates);
+    return std::distance(first, new_last);
+}
+
+template<typename KeyIt, typename ValIt, typename Size>
+void
+generate_data(KeyIt keys_begin, ValIt vals_begin, Size n, std::uint32_t seed)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    if constexpr (std::is_same_v<ValT, Particle>)
+    {
+        static_assert(std::is_same_v<KeyT, Particle::energy_type>);
+        std::default_random_engine gen{seed};
+        std::uniform_real_distribution<float> mass_dist(0.0, 1000.0);
+        std::uniform_real_distribution<float> velocity_dist(0.0, 1.0);
+        std::uniform_real_distribution<float> coord_dist(0.0, 1.0);
+        std::generate(vals_begin, vals_begin + n, [&gen, &mass_dist, &velocity_dist, &coord_dist]() {
+            return Particle{mass_dist(gen), velocity_dist(gen), {coord_dist(gen), coord_dist(gen), coord_dist(gen)}};
+        });
+        std::transform(vals_begin, vals_begin + n, keys_begin, [](const auto& particle) { return particle.energy(); });
+    }
+    else
+    {
+        TestUtils::generate_arithmetic_data(keys_begin, n, seed);
+        TestUtils::generate_arithmetic_data(vals_begin, n, seed + 1);
+        // avoid having value duplicates at the same locations as key duplicates for a more robust stability check
+        std::shuffle(vals_begin, vals_begin + n, std::default_random_engine(seed));
+    }
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, StableSort)
+{
+    if constexpr (std::is_same_v<Compare, DefaultComparator>)
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, UnstableSort)
```

Comment:
Actually, I dont understand why we need that function... I would call the sort algo directly
`oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);`

or if we would like to various stable/unstable yo can define a general caller like that:

auto call_sort = [](auto f, auto&&... args) { f(arg...);}

Reply:
Would you mind if I leave it as is?

It is not obvious to me how using a lambda can further simplify the code. There are 2 `call_sort` overloads, essentially 6-lines each. They are called from 3 functions. It is already pretty minimalistic and simple.

Reply:
Ok, i don't insist